### PR TITLE
Feature/MOB-3944 quote list error handling

### PR DESCRIPTION
**Pull request by AleksanderWedrychowskiKarhoo**
## JIRA
[MOB-3944](https://karhoo.atlassian.net/browse/MOB-3944)

## DESCRIPTION
[Create a description of what you have changed in this branch]

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [x] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [ ] Other: *define what type of change it is*

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [ ] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [ ] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

### CODE REVIEW

File: KarhooUISDK/CommonUI/UITexts.swift [hunk 1]
```diff
@@ -403,6 +403,8 @@ public enum UITexts {
         static let feesAndTaxesIncluded = "Text.Quote.FeesAndTaxesIncluded".localized
         static let result = "Text.Quotes.Result".localized
         static let results = "Text.Quotes.Results".localized
+        static let errorNoAvailabilityForTheRequestTimeTitle = "Text.Quotes.ErrorNoAvailabilityForTheRequestTimeTitle".localized
```
Comment (BartlomiejSopalaKarhoo) [inline KarhooUISDK/CommonUI/UITexts.swift:406]: There will be more errors so maybe create for them separate enum inside Qoutes?

File: KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTablePresenter.swift [hunk 2]
```diff
@@ -20,11 +20,12 @@ class KarhooQuoteListTablePresenter: QuoteListTablePresenter {
     init(
         router: QuoteListTableRouter,
         quotes: [Quote],
+        initialState: QuoteListState? = .loading,
```
Comment (BartlomiejSopalaKarhoo) [inline KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTablePresenter.swift:23]: quotes should be passed in initial state .fetched

Comment (BartlomiejSopalaKarhoo) [reply KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTablePresenter.swift:23]: and initial state should be optional then

Comment (BartlomiejSopalaKarhoo) [reply KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTablePresenter.swift]: you can create some computed value for qoutes based on state

File: KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTable+MVPC.swift [hunk 3]
```diff
@@ -41,7 +42,26 @@ protocol QuoteListTablePresenter: AnyObject {
 
     func updateQuoteListState(_ state: QuoteListState)
 
+    func getErrorViewModel() -> QuoteListTableErrorViewModel
```
Comment (BartlomiejSopalaKarhoo) [inline KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTable+MVPC.swift:45]: We discussed that for all errors should be created some ErrorViewCreator to align single responsibility and open-close principles. Why finally you put it as part of QuoteListTablePresenter ?

File: KarhooUISDK/CommonUI/UITexts.swift (hunk 1 above)
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/CommonUI/UITexts.swift:406]: I'd prefer to keep it this way for now. Maybe we can think about translations reorganization once we'll be preparing the codebase to migrate to common phrase project.

Comment (dianakarh) [reply KarhooUISDK/CommonUI/UITexts.swift:406]: I would add the errors in the UITexts.Errors. Maybe have sth like UITexts.Errors.Quote to give them some separation from the other error types?

File: KarhooUISDK/Screens/QuoteList/QuoteListErrorView.swift [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            options: .curveEaseIn,
+            animations: {
+                viewToAnimate.alpha = UIConstants.Alpha.enabled
+                viewToAnimate.transform = .identity
+            },
+            completion: nil
+        )
+    }
+
+    private func buildSeparator(height: CGFloat? = nil) -> UIView {
```
Comment (BartlomiejSopalaKarhoo) [inline KarhooUISDK/Screens/QuoteList/QuoteListErrorView.swift:124]: should be in separate file to make it reusable in another stacks. Compare to LineView and maybe reorganize them to VarticalLine and HirizontalLine views or something like that.

File: KarhooUISDK/CommonUI/UITexts.swift (hunk 1 above)
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/CommonUI/UITexts.swift:406]: @dianakarh @BartlomiejSopalaKarhoo we should talk about approach, but for now, I'd like to keep it as it is. We may change it later.

**Review by dianakarh — APPROVED** [review]

File: KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift [hunk 5]
```diff
@@ -158,21 +153,19 @@ final class KarhooQuoteListPresenter: QuoteListPresenter {
         }
         
         let noQuotesInSelectedCategory = quotesToShow.isEmpty && fetchedQuotes.all.isEmpty == false
-        let noQuotesForSelectedParapeters = quotesToShow.isEmpty && fetchedQuotes.all.isEmpty && fetchedQuotes.status == .completed
+        let noQuotesForTimeAndArea = fetchedQuotes.all.isEmpty // && fetchedQuotes.status == .completed
```
Comment (BartlomiejSopalaKarhoo) [inline KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift:156]: Why  && .completed condition removed?

**Review by BartlomiejSopalaKarhoo — CHANGES_REQUESTED** [review]
Please create ErrorViewCreator/Factory or something like that for all errors logic to align solid SRP and OCP.

File: KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift [hunk 6]
```diff
@@ -116,7 +111,7 @@ final class KarhooQuoteListPresenter: QuoteListPresenter {
         switch error.type {
         case .noAvailabilityInRequestedArea:
             quoteSearchObservable?.unsubscribe(observer: quotesObserver)
-            onStateUpdated?(.empty(reason: .noAvailabilityInRequestedArea))
+            onStateUpdated?(.empty(reason: .noResults))
```
[the remark this answers is not in the document]
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift:114]: I've changed to error name to simplify the flow.

File: KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift (hunk 5 above)
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift:156]: Because the BE request contains information about area and time. If there are no results in response, we do not need to check if a request is done.

File: KarhooUISDK/CommonUI/UITexts.swift (hunk 1 above)
Comment (BartlomiejSopalaKarhoo) [reply KarhooUISDK/CommonUI/UITexts.swift:406]: 2 : 1 you have been voted down :)

File: KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (BartlomiejSopalaKarhoo) [reply KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift:114]: Hmmm.... for me it's not the same, but let's ask @dianakarh and @mohadian . No results is for me when you try to order taxi in let's say center of London but there is no available driver. noAvailabilityInRequestedArea is when you try to order taxi in Bogószów-Gorce or in other small city, where is no any operating Karhoo fleet.

[the remark this answers is not in the document]
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/Screens/QuoteList/QuoteListPresenter.swift:114]: Ok, after some reconsideration, I think you are right. Update on the go.

File: KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTable+MVPC.swift (hunk 3 above)
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTable+MVPC.swift:45]: Presenter needs to know when the action has been made (click on error control view). We could embed callbacks inside an error case enum, but the result would be opposite → more complex and hard to follow solution.

Comment (BartlomiejSopalaKarhoo) [reply KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTable+MVPC.swift:45]: In mockups there is only one action button to refresh quote. If we will decide to automatically reload quote list I will opt for extracting error logic.

Comment (AleksanderWedrychowskiKarhoo) [reply KarhooUISDK/Screens/QuoteList/QouteListTable/QuoteListTable+MVPC.swift:45]: "Contact Us" is an action as well

**Review by BartlomiejSopalaKarhoo — DISMISSED** [review]

File: KarhooUISDK/CommonUI/UITexts.swift (hunk 1 above)
Comment (dianakarh) [reply KarhooUISDK/CommonUI/UITexts.swift:406]: This is such a small issue that I prefer to move on and get this merged into develop asap. Are we aligned with Android on this (similar key names)? We'll be adopting the Android key names when we merge the 2 projects in Phrase anyway, but until then let's at least keep some consistency with the other platform.

**Review by BartlomiejSopalaKarhoo — APPROVED** [review]

**Review by BartlomiejSopalaKarhoo — APPROVED** [review]

**Review by dianakarh — APPROVED** [review]
